<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="James Ashmore" />


<title>Atlas neighbors</title>

<script src="site_libs/header-attrs-2.6/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">LPMP Analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Processing
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-batch-integration.html">Batch integration</a>
    </li>
    <li>
      <a href="02-reduced-dimensions.html">Dimensionality reduction</a>
    </li>
    <li>
      <a href="03-clustering.html">Clustering</a>
    </li>
    <li>
      <a href="04-marker-detection.html">Marker detection</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="lineage-markers.html">Lineage markers</a>
    </li>
    <li>
      <a href="mouse-integration.html">Mouse integration</a>
    </li>
    <li>
      <a href="human-integration.html">Human integration</a>
    </li>
  </ul>
</li>
<li>
  <a href="interactive.html">Interactive</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/jma1991/LPMP">
    <span class="fab fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Atlas neighbors</h1>
<h4 class="author">James Ashmore</h4>
<h4 class="date">16 March, 2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-03-16
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>LPMP/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201214code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201214)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201214code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201214)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjma1991LPMPtreec70de913777246eab616e682f40f36a1fe980fe5targetblankc70de91a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jma1991/LPMP/tree/c70de913777246eab616e682f40f36a1fe980fe5" target="_blank">c70de91</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjma1991LPMPtreec70de913777246eab616e682f40f36a1fe980fe5targetblankc70de91a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jma1991/LPMP/tree/c70de913777246eab616e682f40f36a1fe980fe5" target="_blank">c70de91</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/cache/
    Ignored:    renv/.gitignore
    Ignored:    renv/library/
    Ignored:    renv/settings.dcf
    Ignored:    shiny/

Untracked files:
    Untracked:  analysis/mouse-neighbors.Rmd
    Untracked:  code/weightedKNN.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="setup" class="section level2">
<h2>Setup</h2>
<p>Setup chunk options:</p>
<pre class="r"><code>knitr::opts_chunk$set(
  autodep    = TRUE,
  cache      = FALSE,
  cache.lazy = FALSE,
  cache.path = &quot;cache/mouse-neighbors.Rmd/&quot;,
  dev        = &quot;png&quot;,
  error      = FALSE,
  message    = FALSE,
  warning    = FALSE,
  fig.align = &quot;center&quot;,
  out.width = &#39;100%&#39;
)</code></pre>
<p>Load Bioconductor packages:</p>
<pre class="r"><code>library(BiocNeighbors)
library(MouseGastrulationData)
library(batchelor)
library(scater)
library(scran)
library(scuttle)</code></pre>
<p>Load CRAN packages:</p>
<pre class="r"><code>library(ggforce)
library(pheatmap)</code></pre>
<p>Source user functions:</p>
<pre class="r"><code>source(&quot;code/splitByCol.R&quot;)
source(&quot;code/weightedKNN.R&quot;)</code></pre>
</div>
<div id="neighbors" class="section level2">
<h2>Neighbors</h2>
<p>Read experiment object:</p>
<pre class="r"><code>sce &lt;- readRDS(&quot;output/mouse-integration.rds&quot;)</code></pre>
<p>Define source annotation:</p>
<pre class="r"><code>sce$source &lt;- &quot;reference&quot;

sce$source[sce$batch == &quot;LPMP_EHF&quot;] &lt;- &quot;query&quot;

sce$source[sce$batch == &quot;LPMP_LHF&quot;] &lt;- &quot;query&quot;</code></pre>
<p>Split experiment object by source:</p>
<pre class="r"><code>sce &lt;- splitByCol(sce, sce$source)</code></pre>
<p>Extract corrected PCA matrix for reference and query data:</p>
<pre class="r"><code>dim &lt;- list(
  reference = reducedDim(sce$reference, &quot;corrected&quot;),
  query = reducedDim(sce$query, &quot;corrected&quot;)
)</code></pre>
<p>Find kNN in reference data for each point in query data:</p>
<pre class="r"><code>set.seed(0011110001)

knn &lt;- queryKNN(X = dim$reference, query = dim$query, k = 20) # k = 20 to match fastMNN</code></pre>
<p>Create matrix of kNN annotations:</p>
<pre class="r"><code>knn$cell &lt;- structure(sce$reference$cell[knn$index], .Dim = dim(knn$index))

knn$stage &lt;- structure(sce$reference$stage[knn$index], .Dim = dim(knn$index))

knn$celltype &lt;- structure(sce$reference$celltype[knn$index], .Dim = dim(knn$index))</code></pre>
<p>Plot UMAP embedding of kNN cells:</p>
<pre class="r"><code>dat &lt;- list(
  reference = makePerCellDF(sce$reference, use.coldata = c(&quot;stage&quot;, &quot;celltype&quot;), use.dimred = &quot;UMAP&quot;),
  query = makePerCellDF(sce$query, use.coldata = &quot;sample&quot;, use.dimred = &quot;UMAP&quot;)
)

dat$reference$neighbor &lt;- &quot;Non-neighbor&quot;

dat$reference$neighbor[sce$reference$cell %in% knn$cell] &lt;- &quot;Neighbor&quot;

dat$reference$neighbor &lt;- factor(dat$reference$neighbor, levels = c(&quot;Neighbor&quot;, &quot;Non-neighbor&quot;))

ggplot(dat$reference, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1) + 
  geom_point(data = dat$query, aes(UMAP.1, UMAP.2, shape = sample), size = 1, inherit.aes = FALSE) + 
  scale_colour_manual(name = &quot;Celltype&quot;, values = EmbryoCelltypeColours) + 
  scale_shape_manual(name = &quot;Query&quot;, values = c(1, 4)) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(neighbor ~ stage, drop = FALSE) + 
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-11-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Plot UMAP embedding of kNN stages:</p>
<pre class="r"><code>dat &lt;- list(
  reference = makePerCellDF(sce$reference, use.coldata = c(&quot;stage&quot;, &quot;celltype&quot;), use.dimred = &quot;UMAP&quot;),
  query = makePerCellDF(sce$query, use.coldata = &quot;sample&quot;, use.dimred = &quot;UMAP&quot;)
)

dat$reference$neighbor &lt;- &quot;Non-neighbor&quot;

dat$reference$neighbor[sce$reference$stage %in% knn$stage] &lt;- &quot;Neighbor&quot;

dat$reference$neighbor &lt;- factor(dat$reference$neighbor, levels = c(&quot;Neighbor&quot;, &quot;Non-neighbor&quot;))

ggplot(dat$reference, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1) + 
  geom_point(data = dat$query, aes(UMAP.1, UMAP.2, shape = sample), size = 1, inherit.aes = FALSE) + 
  scale_colour_manual(name = &quot;Celltype&quot;, values = EmbryoCelltypeColours) + 
  scale_shape_manual(name = &quot;Query&quot;, values = c(1, 4)) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(neighbor ~ stage, drop = FALSE) + 
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-12-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Plot UMAP embedding of kNN celltypes:</p>
<pre class="r"><code>dat &lt;- list(
  reference = makePerCellDF(sce$reference, use.coldata = c(&quot;stage&quot;, &quot;celltype&quot;), use.dimred = &quot;UMAP&quot;),
  query = makePerCellDF(sce$query, use.coldata = &quot;sample&quot;, use.dimred = &quot;UMAP&quot;)
)

dat$reference$neighbor &lt;- &quot;Non-neighbor&quot;

dat$reference$neighbor[dat$reference$celltype %in% knn$celltype] &lt;- &quot;Neighbor&quot;

dat$reference$neighbor &lt;- factor(dat$reference$neighbor, levels = c(&quot;Neighbor&quot;, &quot;Non-neighbor&quot;))

ggplot(dat$reference, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1) + 
  geom_point(data = dat$query, aes(UMAP.1, UMAP.2, shape = sample), size = 1, inherit.aes = FALSE) + 
  scale_colour_manual(name = &quot;Celltype&quot;, values = EmbryoCelltypeColours) + 
  scale_shape_manual(name = &quot;Query&quot;, values = c(1, 4)) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(neighbor ~ stage, drop = FALSE) +  
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-13-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Annotate query data with weighted kNN annotations:</p>
<pre class="r"><code>sce$query$cell.mapped &lt;- as.list(as.data.frame(t(knn$cell)))

sce$query$stage.mapped &lt;- weightedKNN(knn, class = &quot;stage&quot;)

sce$query$celltype.mapped &lt;- weightedKNN(knn, class = &quot;celltype&quot;)</code></pre>
<p>Self-annotate reference data:</p>
<pre class="r"><code>sce$reference$cell.mapped &lt;- sce$reference$cell

sce$reference$stage.mapped &lt;- sce$reference$stage

sce$reference$celltype.mapped &lt;- sce$reference$celltype</code></pre>
<p>Subset reference data to kNN annotations:</p>
<pre class="r"><code>id1 &lt;- sce$reference$celltype %in% sce$query$celltype.mapped

id2 &lt;- sce$reference$stage %in% sce$query$stage.mapped

idx &lt;- id1 &amp; id2

sce$reference &lt;- sce$reference[, idx]</code></pre>
<p>Combine reference and query data:</p>
<pre class="r"><code>sce &lt;- cbind(sce$reference, sce$query)</code></pre>
</div>
<div id="integration" class="section level2">
<h2>Integration</h2>
<div id="variance-modelling" class="section level3">
<h3>Variance modelling</h3>
<p>Model the per-gene variance:</p>
<pre class="r"><code>dec &lt;- modelGeneVar(sce, block = sce$sample)</code></pre>
</div>
<div id="feature-selection" class="section level3">
<h3>Feature selection</h3>
<p>Read blacklist genes and subset by category:</p>
<pre class="r"><code>nan &lt;- read.csv(&quot;data/blacklist.csv&quot;)

nan &lt;- subset(nan, CATEGORY %in% c(&quot;CC&quot;, &quot;MT&quot;, &quot;RP&quot;, &quot;XIST&quot;, &quot;Y&quot;))</code></pre>
<p>Ignore blacklist genes in feature selection:</p>
<pre class="r"><code>dec$blacklist &lt;- FALSE

dec$blacklist[rownames(dec) %in% nan$ENSEMBL] &lt;- TRUE</code></pre>
<p>Identify highly variable genes:</p>
<pre class="r"><code>hvg &lt;- subset(dec, blacklist == FALSE)

hvg &lt;- getTopHVGs(hvg, var.field = &quot;bio&quot;, var.threshold = 0, fdr.field = &quot;p.value&quot;, fdr.threshold = 0.05)</code></pre>
</div>
<div id="batch-normalization" class="section level3">
<h3>Batch normalization</h3>
<p>Perform scaling normalization:</p>
<pre class="r"><code>sce &lt;- multiBatchNorm(sce, batch = sce$sample)</code></pre>
</div>
<div id="batch-integration" class="section level3">
<h3>Batch integration</h3>
<p>Create sequential merge order:</p>
<pre class="r"><code>tab &lt;- as.data.frame(table(sce$sample, sce$stage), stringsAsFactors = FALSE)

colnames(tab) &lt;- c(&quot;sample&quot;, &quot;stage&quot;, &quot;ncells&quot;)

tab &lt;- subset(tab, ncells &gt; 0)

tab &lt;- subset(tab, sample %in% AtlasSampleMetadata$sample)

lvl &lt;- c(&quot;E6.5&quot;, &quot;E6.75&quot;, &quot;E7.0&quot;, &quot;mixed_gastrulation&quot;, &quot;E7.25&quot;, &quot;E7.5&quot;, &quot;E7.75&quot;, &quot;E8.0&quot;, &quot;E8.25&quot;, &quot;E8.5&quot;)

tab$stage &lt;- factor(tab$stage, levels = lvl)

tab &lt;- tab[order(tab$stage, tab$ncells, decreasing = TRUE), ]

idx &lt;- c(tab$sample, &quot;LPMP_LHF&quot;, &quot;LPMP_EHF&quot;)</code></pre>
<p>Correct experiments using nearest neighbours:</p>
<pre class="r"><code>set.seed(0000110010)

mnn &lt;- correctExperiments(
  sce,
  batch = sce$sample,
  subset.row = hvg,
  correct.all = TRUE,
  PARAM = FastMnnParam(merge.order = idx)
)</code></pre>
<p>Store combined variance model:</p>
<pre class="r"><code>metadata(mnn)$modelGeneVar &lt;- dec</code></pre>
<p>Store highly variable genes:</p>
<pre class="r"><code>metadata(mnn)$getTopHVGs &lt;- hvg</code></pre>
</div>
</div>
<div id="diagnostics" class="section level2">
<h2>Diagnostics</h2>
<p>Check proportion of lost variance:</p>
<pre class="r"><code>var &lt;- metadata(mnn)$merge.info$lost.var

col &lt;- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = &quot;Reds&quot;))(100)

brk &lt;- seq(0, max(var), length.out = 101)

pheatmap(var[, idx], color = col, display_numbers = TRUE, number_format = &quot;%.2f&quot;, cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-27-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Cluster on the corrected PCA to obtain a partitioning of the cells:</p>
<pre class="r"><code>snn &lt;- buildSNNGraph(mnn, type = &quot;jaccard&quot;, use.dimred = &quot;corrected&quot;)

com &lt;- igraph::cluster_louvain(snn)</code></pre>
<p>Tabulate number of cells from multiple batches within each cluster:</p>
<pre class="r"><code>tab &lt;- table(Cluster = com$membership, Batch = mnn$batch)

col &lt;- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = &quot;Blues&quot;))(100)

brk &lt;- seq(0, max(tab), length.out = 101)

pheatmap(tab[, idx], color = col, display_numbers = TRUE, number_format = &quot;%d&quot;, cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-29-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Perform TSNE on the corrected PCA matrix:</p>
<pre class="r"><code>set.seed(42)

mnn &lt;- runTSNE(mnn, dimred = &quot;corrected&quot;, perplexity = 120) # same settings as reference analysis</code></pre>
<p>Visualize the corrected PCA using a TSNE plot:</p>
<pre class="r"><code>ggcells(mnn, aes(TSNE.1, TSNE.2, colour = celltype.mapped)) + 
  geom_point(size = 0.1, show.legend = FALSE) + 
  scale_colour_manual(values = EmbryoCelltypeColours, na.value = &quot;#000000&quot;) + 
  labs(x = &quot;TSNE 1&quot;, y = &quot;TSNE 2&quot;) + 
  facet_wrap(~ sample) + 
  theme_bw() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-31-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Perform UMAP on the corrected PCA matrix:</p>
<pre class="r"><code>set.seed(42)

mnn &lt;- runUMAP(mnn, dimred = &quot;corrected&quot;, n_neighbors = 20, min_dist = 0.7) # same settings as reference analysis</code></pre>
<p>Visualize the corrected PCA using a UMAP plot:</p>
<pre class="r"><code>ggcells(mnn, aes(UMAP.1, UMAP.2, colour = celltype.mapped)) + 
  geom_point(size = 0.1, show.legend = FALSE) + 
  scale_colour_manual(values = EmbryoCelltypeColours, na.value = &quot;#000000&quot;) + 
  labs(x = &quot;UMAP 1&quot;, y = &quot;UMAP 2&quot;) + 
  facet_wrap(~ sample) + 
  theme_bw() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-33-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Plot TSNE embedding of mapped stage and mapped celltype:</p>
<pre class="r"><code>dat &lt;- list(
  all = makePerCellDF(mnn, use.coldata = NULL, use.dimred = &quot;TSNE&quot;),
  col = makePerCellDF(mnn, use.coldata = c(&quot;stage.mapped&quot;, &quot;celltype.mapped&quot;, &quot;source&quot;, &quot;sample&quot;), use.dimred = &quot;TSNE&quot;)
)

dat$col$source &lt;- &quot;Embryo Atlas&quot;

dat$col$source[dat$col$sample == &quot;LPMP_EHF&quot;] &lt;- &quot;LPMP EHF&quot;

dat$col$source[dat$col$sample == &quot;LPMP_LHF&quot;] &lt;- &quot;LPMP LHF&quot;

dat$col$source &lt;- factor(dat$col$source, levels = c(&quot;Embryo Atlas&quot;, &quot;LPMP LHF&quot;, &quot;LPMP EHF&quot;))

ggplot() + 
  geom_point(data = dat$all, aes(TSNE.1, TSNE.2), colour = &quot;grey92&quot;, size = 0.1) + 
  geom_point(data = dat$col, aes(TSNE.1, TSNE.2, colour = celltype.mapped), size = 0.1) + 
  scale_colour_manual(name = &quot;Celltype (Mapped)&quot;, values = EmbryoCelltypeColours) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(source ~ stage.mapped) +  
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-34-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Plot UMAP embedding of mapped stage and mapped celltype:</p>
<pre class="r"><code>dat &lt;- list(
  all = makePerCellDF(mnn, use.coldata = NULL, use.dimred = &quot;UMAP&quot;),
  col = makePerCellDF(mnn, use.coldata = c(&quot;stage.mapped&quot;, &quot;celltype.mapped&quot;, &quot;source&quot;, &quot;sample&quot;), use.dimred = &quot;UMAP&quot;)
)

dat$col$source &lt;- &quot;Embryo Atlas&quot;

dat$col$source[dat$col$sample == &quot;LPMP_EHF&quot;] &lt;- &quot;LPMP EHF&quot;

dat$col$source[dat$col$sample == &quot;LPMP_LHF&quot;] &lt;- &quot;LPMP LHF&quot;

dat$col$source &lt;- factor(dat$col$source, levels = c(&quot;Embryo Atlas&quot;, &quot;LPMP LHF&quot;, &quot;LPMP EHF&quot;))

ggplot() + 
  geom_point(data = dat$all, aes(UMAP.1, UMAP.2), colour = &quot;grey92&quot;, size = 0.1) + 
  geom_point(data = dat$col, aes(UMAP.1, UMAP.2, colour = celltype.mapped), size = 0.1) + 
  scale_colour_manual(name = &quot;Celltype (Mapped)&quot;, values = EmbryoCelltypeColours) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(source ~ stage.mapped) + 
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-35-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="annotation" class="section level2">
<h2>Annotation</h2>
<p>Split experiment object by source:</p>
<pre class="r"><code>mnn &lt;- splitByCol(mnn, mnn$source)</code></pre>
<p>Extract corrected PCA matrix for reference and query data:</p>
<pre class="r"><code>dim &lt;- list(
  reference = reducedDim(mnn$reference, &quot;corrected&quot;),
  query = reducedDim(mnn$query, &quot;corrected&quot;)
)</code></pre>
<p>Find kNN in reference data for each point in query data:</p>
<pre class="r"><code>set.seed(0011110001)

knn &lt;- queryKNN(X = dim$reference, query = dim$query, k = 20) # k = 20 to match fastMNN</code></pre>
<p>Create matrix of kNN annotations:</p>
<pre class="r"><code>knn$cell &lt;- structure(mnn$reference$cell[knn$index], .Dim = dim(knn$index))

knn$stage &lt;- structure(mnn$reference$stage[knn$index], .Dim = dim(knn$index))

knn$celltype &lt;- structure(mnn$reference$celltype[knn$index], .Dim = dim(knn$index))</code></pre>
<p>Plot UMAP embedding of kNN cells:</p>
<pre class="r"><code>dat &lt;- list(
  reference = makePerCellDF(mnn$reference, use.coldata = c(&quot;stage&quot;, &quot;celltype&quot;), use.dimred = &quot;UMAP&quot;),
  query = makePerCellDF(mnn$query, use.coldata = &quot;sample&quot;, use.dimred = &quot;UMAP&quot;)
)

dat$reference$neighbor &lt;- &quot;Non-neighbor&quot;

dat$reference$neighbor[mnn$reference$cell %in% knn$cell] &lt;- &quot;Neighbor&quot;

dat$reference$neighbor &lt;- factor(dat$reference$neighbor, levels = c(&quot;Neighbor&quot;, &quot;Non-neighbor&quot;))

ggplot(dat$reference, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1) + 
  geom_point(data = dat$query, aes(UMAP.1, UMAP.2, shape = sample), size = 1, inherit.aes = FALSE) + 
  scale_colour_manual(name = &quot;Celltype&quot;, values = EmbryoCelltypeColours) + 
  scale_shape_manual(name = &quot;Query&quot;, values = c(1, 4)) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(neighbor ~ stage, drop = FALSE) + 
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-40-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Plot UMAP embedding of kNN stages:</p>
<pre class="r"><code>dat &lt;- list(
  reference = makePerCellDF(mnn$reference, use.coldata = c(&quot;stage&quot;, &quot;celltype&quot;), use.dimred = &quot;UMAP&quot;),
  query = makePerCellDF(mnn$query, use.coldata = &quot;sample&quot;, use.dimred = &quot;UMAP&quot;)
)

dat$reference$neighbor &lt;- &quot;Non-neighbor&quot;

dat$reference$neighbor[mnn$reference$stage %in% knn$stage] &lt;- &quot;Neighbor&quot;

dat$reference$neighbor &lt;- factor(dat$reference$neighbor, levels = c(&quot;Neighbor&quot;, &quot;Non-neighbor&quot;))

ggplot(dat$reference, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1) + 
  geom_point(data = dat$query, aes(UMAP.1, UMAP.2, shape = sample), size = 1, inherit.aes = FALSE) + 
  scale_colour_manual(name = &quot;Celltype&quot;, values = EmbryoCelltypeColours) + 
  scale_shape_manual(name = &quot;Query&quot;, values = c(1, 4)) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(neighbor ~ stage, drop = FALSE) + 
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-41-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Plot UMAP embedding of kNN celltypes:</p>
<pre class="r"><code>dat &lt;- list(
  reference = makePerCellDF(mnn$reference, use.coldata = c(&quot;stage&quot;, &quot;celltype&quot;), use.dimred = &quot;UMAP&quot;),
  query = makePerCellDF(mnn$query, use.coldata = &quot;sample&quot;, use.dimred = &quot;UMAP&quot;)
)

dat$reference$neighbor &lt;- &quot;Non-neighbor&quot;

dat$reference$neighbor[dat$reference$celltype %in% knn$celltype] &lt;- &quot;Neighbor&quot;

dat$reference$neighbor &lt;- factor(dat$reference$neighbor, levels = c(&quot;Neighbor&quot;, &quot;Non-neighbor&quot;))

ggplot(dat$reference, aes(UMAP.1, UMAP.2, colour = celltype)) + 
  geom_point(size = 0.1) + 
  geom_point(data = dat$query, aes(UMAP.1, UMAP.2, shape = sample), size = 1, inherit.aes = FALSE) + 
  scale_colour_manual(name = &quot;Celltype&quot;, values = EmbryoCelltypeColours) + 
  scale_shape_manual(name = &quot;Query&quot;, values = c(1, 4)) + 
  guides(colour = guide_legend(override.aes = list(size = 1))) + 
  facet_grid(neighbor ~ stage, drop = FALSE) +  
  theme_no_axes() + 
  theme(aspect.ratio = 1)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-42-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Annotate query data with weighted kNN annotations:</p>
<pre class="r"><code>mnn$query$cell.remapped &lt;- as.list(as.data.frame(t(knn$cell)))

mnn$query$stage.remapped &lt;- weightedKNN(knn, class = &quot;stage&quot;)

mnn$query$celltype.remapped &lt;- weightedKNN(knn, class = &quot;celltype&quot;)</code></pre>
<p>Self-annotate reference data:</p>
<pre class="r"><code>mnn$reference$cell.remapped &lt;- mnn$reference$cell

mnn$reference$stage.remapped &lt;- mnn$reference$stage

mnn$reference$celltype.remapped &lt;- mnn$reference$celltype</code></pre>
<p>Compare mapped and remapped stage annotations:</p>
<pre class="r"><code>tab &lt;- table(Mapped = mnn$query$stage.mapped, Remapped = mnn$query$stage.remapped)

col &lt;- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = &quot;Blues&quot;))(100)

brk &lt;- seq(0, max(tab), length.out = 101)

pheatmap(tab, color = col, display_numbers = TRUE, number_format = &quot;%d&quot;, cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 0)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-45-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Compare mapped and remapped celltype annotations:</p>
<pre class="r"><code>tab &lt;- table(Mapped = mnn$query$celltype.mapped, Remapped = mnn$query$celltype.remapped)

col &lt;- colorRampPalette(RColorBrewer::brewer.pal(n = 5, name = &quot;Blues&quot;))(100)

brk &lt;- seq(0, max(tab), length.out = 101)

pheatmap(tab, color = col, display_numbers = TRUE, number_format = &quot;%d&quot;, cluster_cols = FALSE, cluster_rows = FALSE, angle_col = 90)</code></pre>
<p><img src="figure/mouse-neighbors.Rmd/unnamed-chunk-46-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Combine reference and query data:</p>
<pre class="r"><code>mnn &lt;- cbind(mnn$reference, mnn$query)</code></pre>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<div id="output" class="section level3">
<h3>Output</h3>
<p>Save experiment object to disk:</p>
<pre class="r"><code>saveRDS(mnn, file = &quot;output/mouse-neighbors.rds&quot;)</code></pre>
</div>
<div id="session" class="section level3">
<h3>Session</h3>
<p>Print session information:</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.2 (2020-06-22)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices datasets  utils    
[8] methods   base     

other attached packages:
 [1] pheatmap_1.0.12             ggforce_0.3.2              
 [3] scuttle_1.0.4               scran_1.18.3               
 [5] scater_1.18.3               ggplot2_3.3.3              
 [7] batchelor_1.6.2             MouseGastrulationData_1.4.0
 [9] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
[11] Biobase_2.50.0              GenomicRanges_1.42.0       
[13] GenomeInfoDb_1.26.2         IRanges_2.24.1             
[15] S4Vectors_0.28.1            BiocGenerics_0.36.0        
[17] MatrixGenerics_1.2.0        matrixStats_0.57.0         
[19] BiocNeighbors_1.8.2         workflowr_1.6.2            

loaded via a namespace (and not attached):
  [1] Rtsne_0.15                    ggbeeswarm_0.6.0             
  [3] colorspace_2.0-0              ellipsis_0.3.1               
  [5] rprojroot_2.0.2               bluster_1.0.0                
  [7] XVector_0.30.0                fs_1.5.0                     
  [9] rstudioapi_0.13               farver_2.0.3                 
 [11] bit64_4.0.5                   RSpectra_0.16-0              
 [13] interactiveDisplayBase_1.28.0 AnnotationDbi_1.52.0         
 [15] codetools_0.2-18              sparseMatrixStats_1.2.0      
 [17] cachem_1.0.1                  knitr_1.30                   
 [19] polyclip_1.10-0               ResidualMatrix_1.0.0         
 [21] dbplyr_2.0.0                  uwot_0.1.10                  
 [23] shiny_1.6.0                   BiocManager_1.30.10          
 [25] compiler_4.0.2                httr_1.4.2                   
 [27] dqrng_0.2.1                   assertthat_0.2.1             
 [29] Matrix_1.3-2                  fastmap_1.1.0                
 [31] limma_3.46.0                  later_1.1.0.1                
 [33] tweenr_1.0.1                  BiocSingular_1.6.0           
 [35] htmltools_0.5.1.1             tools_4.0.2                  
 [37] rsvd_1.0.3                    igraph_1.2.6                 
 [39] gtable_0.3.0                  glue_1.4.2                   
 [41] GenomeInfoDbData_1.2.4        dplyr_1.0.3                  
 [43] rappdirs_0.3.1                Rcpp_1.0.6                   
 [45] vctrs_0.3.6                   ExperimentHub_1.16.0         
 [47] DelayedMatrixStats_1.12.2     xfun_0.20                    
 [49] stringr_1.4.0                 beachmat_2.6.4               
 [51] mime_0.9                      lifecycle_0.2.0              
 [53] irlba_2.3.3                   renv_0.12.5                  
 [55] statmod_1.4.35                AnnotationHub_2.22.0         
 [57] edgeR_3.32.1                  zlibbioc_1.36.0              
 [59] MASS_7.3-53                   scales_1.1.1                 
 [61] promises_1.1.1                RColorBrewer_1.1-2           
 [63] yaml_2.2.1                    curl_4.3                     
 [65] memoise_2.0.0                 gridExtra_2.3                
 [67] stringi_1.5.3                 RSQLite_2.2.3                
 [69] BiocVersion_3.12.0            BiocParallel_1.24.1          
 [71] rlang_0.4.10                  pkgconfig_2.0.3              
 [73] bitops_1.0-6                  evaluate_0.14                
 [75] lattice_0.20-41               purrr_0.3.4                  
 [77] labeling_0.4.2                bit_4.0.4                    
 [79] tidyselect_1.1.0              RcppAnnoy_0.0.18             
 [81] magrittr_2.0.1                R6_2.5.0                     
 [83] generics_0.1.0                DelayedArray_0.16.1          
 [85] DBI_1.1.1                     pillar_1.4.7                 
 [87] withr_2.4.1                   RCurl_1.98-1.2               
 [89] tibble_3.0.5                  crayon_1.3.4                 
 [91] BiocFileCache_1.14.0          rmarkdown_2.6                
 [93] viridis_0.5.1                 locfit_1.5-9.4               
 [95] grid_4.0.2                    blob_1.2.1                   
 [97] git2r_0.28.0                  digest_0.6.27                
 [99] xtable_1.8-4                  httpuv_1.5.5                 
[101] munsell_0.5.0                 beeswarm_0.2.3               
[103] viridisLite_0.3.0             vipor_0.4.5                  </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
